hypothesis_regex.py

```diff
@@ -1,102 +1,325 @@
+            if context.flags & re.IGNORECASE:
+                return hs.sampled_from([c.lower(), c.upper()])
+            else:
+                return hs.just(c)
+
         elif code == sre.NOT_LITERAL:
-            return hs.characters(blacklist_characters=[chr(value)])
+            # Regex '[^a]' (negation of a single char)
+            c = six.unichr(value)
+            blacklist = set([c.lower(), c.upper()]) \
+                if context.flags & re.IGNORECASE else [c]
+            return hs.characters(blacklist_characters=blacklist)
+
         elif code == sre.IN:
+            # Regex '[abc0-9]' (set of characters)
             charsets = value
 
-            chars = []
+            builder = CharactersBuilder(negate=charsets[0][0] == sre.NEGATE,
+                                        flags=context.flags)
+
             for charset_code, charset_value in charsets:
                 if charset_code == sre.NEGATE:
+                    # Regex '[^...]' (negation)
                     pass
                 elif charset_code == sre.LITERAL:
-                    chars.append(chr(charset_value))
+                    # Regex '[a]' (single char)
+                    builder.add_chars(six.unichr(charset_value))
                 elif charset_code == sre.RANGE:
+                    # Regex '[a-z]' (char range)
                     low, high = charset_value
-                    chars.extend([
-                        chr(x) for x in six.moves.range(low, high+1)
-                    ])
+                    for x in six.moves.range(low, high+1):
+                        builder.add_chars(six.unichr(x))
                 elif charset_code == sre.CATEGORY:
-                    if charset_value not in _CATEGORIES:
-                        raise NotImplementedError(
-                            'Unknown char category: %s' % charset_value
-                        )
-                    chars.extend(_CATEGORIES[charset_value])
+                    # Regex '[\w]' (char category)
+                    builder.add_category(charset_value)
                 else:
-                    raise NotImplementedError(
+                    raise he.InvalidArgument(
                         'Unknown charset code: %s' % charset_code
                     )
 
-            if charsets[0][0] == sre.NEGATE:
-                return hs.characters(blacklist_characters=chars)
-            else:
-                return hs.text(alphabet=chars, min_size=1, max_size=1)
+            return builder.strategy
 
         elif code == sre.ANY:
-            # TODO: consider checking MULTILINE flag
-            return hs.characters(blacklist_characters="\n")
+            # Regex '.' (any char)
+            if context.flags & re.DOTALL:
+                return hs.characters()
+            else:
+                return hs.characters(blacklist_characters="\n")
+
         elif code == sre.AT:
+            # Regexes like '^...', '...$', '\bfoo', '\Bfoo'
             return hs.just('')
```

Comment:
It doesn't mean that it's not working. Usually people place "^" at the beginning of their regexes and "$" - at the end. And in that case it will work just fine. I also think about cases where "^" or "$" are in the middle and MULTILINE is enabled, so that those chars can generate newlines.

Also, "\b" (word boundary) and "\B" (not a word boundary) might work.

To enforce those invariants, I'm considering of adding a filter to resulting strategy that will check that generated example actually matches regex. What do you think about that?